internal/pkg/print/debug.go

```diff
@@ -57,6 +81,147 @@ func BuildDebugStrFromSlice(inputSlice []string) string {
+	}
+	var bodyMap map[string]any
+	if len(body) != 0 {
+		if err := json.Unmarshal(body, &bodyMap); err != nil {
+			return []string{status, headers}, fmt.Errorf("unmarshal request body: %w", err)
+		}
+	}
+
+	// restore body
+	req.Body = io.NopCloser(strings.NewReader(string(body)))
+	payload := fmt.Sprintf("request body: %v", BuildDebugStrFromMap(bodyMap))
+
+	return []string{status, headers, payload}, nil
+}
+
+// BuildDebugStrFromHTTPResponse converts an HTTP response to a user-friendly string representation.
+// This function also receives a list of headers to include in the output, if empty, the default headers are used.
+// The return value is a list of strings that should be printed separately.
+func BuildDebugStrFromHTTPResponse(resp *http.Response, includeHeaders []string) ([]string, error) {
+	if resp == nil {
+		return nil, fmt.Errorf("response is nil")
+	}
+
+	if resp.Request == nil || resp.Proto == "" || resp.Status == "" {
+		return nil, fmt.Errorf("response is invalid")
+	}
+
+	status := fmt.Sprintf("response from %s: %s %s", resp.Request.URL, resp.Proto, resp.Status)
+
+	headersMap := buildHeaderMap(resp.Header, includeHeaders)
+	headers := fmt.Sprintf("response headers: %v", BuildDebugStrFromMap(headersMap))
+
+	if resp.Body == nil {
+		return []string{status, headers}, nil
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return []string{status, headers}, fmt.Errorf("read response body: %w", err)
+	}
+	var bodyMap map[string]any
+	if len(body) != 0 {
+		if err := json.Unmarshal(body, &bodyMap); err != nil {
+			return []string{status, headers}, fmt.Errorf("unmarshal response body: %w", err)
+		}
+	}
+
+	// restore body
+	resp.Body = io.NopCloser(strings.NewReader(string(body)))
+	payload := fmt.Sprintf("response body: %v", BuildDebugStrFromMap(bodyMap))
+
+	return []string{status, headers, payload}, nil
+}
+
+// RequestResponseCapturer is a middleware that captures the request and response of an HTTP request.
+// Receives a printer and a list of headers to include in the output
+// If the list of headers is empty, the default headers are used.
+// The printer is used to print the captured data.
+func RequestResponseCapturer(p *Printer, includeHeaders []string) config.Middleware {
+	return func(rt http.RoundTripper) http.RoundTripper {
+		return &roundTripperWithCapture{rt, p, includeHeaders}
+	}
+}
+
+type roundTripperWithCapture struct {
+	transport    http.RoundTripper
+	p            *Printer
+	debugHeaders []string
+}
+
+func (rt roundTripperWithCapture) RoundTrip(req *http.Request) (*http.Response, error) {
+	reqStr, err := BuildDebugStrFromHTTPRequest(req, rt.debugHeaders)
+	if err != nil {
+		rt.p.Debug(ErrorLevel, "printing request to debug logs: %v", err)
+	}
+	for _, line := range reqStr {
+		rt.p.Debug(DebugLevel, line)
+	}
+	resp, err := rt.transport.RoundTrip(req)
+	defer func() {
+		if err == nil {
+			respStrSlice, err := BuildDebugStrFromHTTPResponse(resp, rt.debugHeaders)
```

Comment:
```suggestion
			respStrSlice, tempErr := BuildDebugStrFromHTTPResponse(resp, rt.debugHeaders)
```
To not overwrite the main function error